semantic-conventions/src/opentelemetry/semconv/templating/compatibility.py

```diff
@@ -0,0 +1,191 @@
+from opentelemetry.semconv.model.semantic_attribute import (
+    EnumAttributeType,
+    EnumMember,
+    RequirementLevel,
+    SemanticAttribute,
+    StabilityLevel,
+)
+from opentelemetry.semconv.model.semantic_convention import (
+    MetricSemanticConvention,
+    SemanticConventionSet,
+)
+
+
+class Problem:
+    signal: str
+    name: str
+    message: str
+    critical: bool
+
+    def __init__(self, signal: str, name: str, message: str, critical: bool = True):
+        self.signal = signal
+        self.name = name
+        self.message = message
+        self.critical = critical
+
+    def __str__(self):
+        return f"\t {self.signal} '{self.name}': {self.message}"
+
+    def __eq__(self, other):
+        if isinstance(other, self.__class__):
+            return self.__dict__ == other.__dict__
+
+        return False
+
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+
+class CompatibilityChecker:
+    previous_version: str
+
+    def __init__(
+        self,
+        current_semconv: SemanticConventionSet,
+        previous_semconv: SemanticConventionSet,
+    ):
+        self.current_semconv = current_semconv
+        self.previous_semconv = previous_semconv
+
+    def check(self) -> list[Problem]:
+        problems = []  # type: list[Problem]
+        for semconv in self.previous_semconv.models.values():
+            for prev_attr in semconv.attributes_and_templates:
+                if (
+                    prev_attr.is_local
+                    and prev_attr.attr_id is not None
+                    and prev_attr.ref is None
+                ):
+                    self._check_attribute(prev_attr, problems)
+            if isinstance(semconv, MetricSemanticConvention):
+                self._check_metric(semconv, problems)
+        return problems
+
+    def _check_attribute(self, prev: SemanticAttribute, problems: list[Problem]):
+        cur = self.current_semconv._lookup_attribute(prev.fqn)
+        if cur is None:
+            problems.append(Problem("attribute", prev.fqn, "was removed"))
+            return
+
+        if prev.stability == StabilityLevel.STABLE:
+            if cur.stability != prev.stability:
+                problems.append(
+                    Problem(
+                        "attribute",
+                        prev.fqn,
+                        f"stability changed from '{prev.stability}' to '{cur.stability}'",
+                    )
+                )
+
+            if isinstance(prev.attr_type, EnumAttributeType):
```

Comment:
Should we prevent *any* type change on a stable attribute?

Reply:
we do (line 102), it just needs to happen in different ways for enums and non-enums